### ISSUE TITLE
x/tools/cmd/gopls: signatureHelp should return position info for signatures

### ISSUE THREAD

**Issue by myitcv** [golang/go#32468]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version devel +385b2e0cac Fri May 24 21:34:53 2019 +0000 linux/amd64
$ go list -m golang.org/x/tools
golang.org/x/tools v0.0.0-20190521203540-521d6ed310dd
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN="/home/myitcv/gostuff/src/github.com/myitcv/govim/cmd/govim/.bin"
GOCACHE="/home/myitcv/.cache/go-build"
GOENV="/home/myitcv/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/myitcv/gostuff"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/myitcv/gos"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/myitcv/gos/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/myitcv/gostuff/src/github.com/myitcv/govim/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build579955727=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

The signature help call from client to server (https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md#signature-help-request-leftwards_arrow_with_hook) returns a list of signatures:

```go
type SignatureHelp struct {
	Signatures []SignatureInformation `json:"signatures"`
	ActiveSignature float64 `json:"activeSignature"`
	ActiveParameter float64 `json:"activeParameter"`
}

type SignatureInformation struct {
	Label string `json:"label"`
	Documentation string `json:"documentation,omitempty"` 
	Parameters []ParameterInformation `json:"parameters,omitempty"`
}

type ParameterInformation struct {
	Label string `json:"label"` 
	Documentation string `json:"documentation,omitempty"` 
}
```

But what's strange is that the signature information does not include position information, i.e. the start (and end, where it's valid/available) of where the call signature is valid: think the starting `(` (to the end `)`).

This position information is valuable to my mind because it tells the client where to place, for example, a popup window.

---

cc @stamblerre @ianthehat

**Comment by stamblerre** [comment]
According to the LSP spec, the ParameterInformation struct can include position information instead of a label. If we used that, would it be enough to solve this problem?

**Comment by myitcv** [comment]
Oh wow, I totally missed that comment, thanks.

Yes, if this is position information like other places in the spec, should be fine!

**Comment by stamblerre** [comment]
@pjweinb: Would it be possible for us to make this change in the protocol generator?

**Comment by pjweinb** [comment]
I'll look at it when I get back, but seems likely.

**Comment by pjweinb** [comment]
@stamblerre: could you add a [2]float64 field to ParameterInformation in lsp/source/signature_help.go? It does't matter what you name it, but once it exists (and gets filled in), we can change line 57 in signature_help.go to use the new field in place of p.Label.

**Comment by myitcv** [comment]
@stamblerre @pjweinb - is there anything I can do to help move this along?

**Comment by stamblerre** [comment]
I'm sorry, this is a slightly lower priority issue for us right now. I will have to make the changes Peter requested before we can change the generator, but I haven't had a chance to get to this yet.

**Comment by oblitum** [comment]
For what is worth, I guess it's the same issue, this was fixed on the bingo server long ago https://github.com/saibing/bingo/issues/20.

**Comment by oblitum** [comment]
In specific there was a mention to `labelOffsetSupport`: https://github.com/saibing/bingo/issues/20#issuecomment-450676103.

**Comment by oblitum** [comment]
And the advent of the gopls server at the time caused a regression for bingo: https://github.com/saibing/bingo/issues/135#issuecomment-482658628.

**Comment by stamblerre** [comment]
Thanks for reporting this. It's a slightly different issue because it's a bug in signature help, whereas this issue is to track making use of the `labelOffsetSupport` capability, as you said. I filed https://github.com/golang/go/issues/36766 to track the bug you mentioned.
